tensorboard/program.py

```diff
@@ -218,6 +227,38 @@ def main(self, ignored_argv=('',)):
       sys.stderr.flush()
       return -1
 
+  def _install_sigterm_handler(self):
+    old_sigterm_handler = None  # set below
+    def handle_sigterm(signal_number, stack_frame):
+      del signal_number  # unused
+      del stack_frame  # unused
+      # In case we get SIGTERMed again while running atexit handlers,
+      # take the hint and actually die.
+      signal.signal(signal.SIGTERM, old_sigterm_handler)
+      sys.stderr.write("TensorBoard caught SIGTERM; exiting.\n")
+      # The main thread is the only non-daemon thread, so it suffices to
+      # exit hence.
+      sys.exit(0)
+    old_sigterm_handler = signal.signal(signal.SIGTERM, handle_sigterm)
+
+  def _register_info(self, server):
+    server_url = urllib.parse.urlparse(server.get_url())
+    info = manager.TensorboardInfo(
+        version=version.VERSION,
+        start_time=datetime.datetime.now(),
+        port=server_url.port,
+        pid=os.getpid(),
+        path_prefix=self.flags.path_prefix,
+        logdir=self.flags.logdir,
+        db=self.flags.db,
+        cache_key=manager.cache_key(
+            working_directory=os.getcwd(),
+            arguments=sys.argv[1:],
```

Comment:
Hmm, good point. I’ll have to think about the best way to do this,
particularly given the argparse/abseil duality that already exists in
`configure` (which may or may not actually be a problem; it’s just not
obvious to me what it does in all different environments).

What would you think about using `kwargs` to change the input `argv`
before parsing it, rather than using fixing up the `flags` object
afterward? The implementation is a bit messier, but the API is the same,
and it guarantees that there’s no skew in the “actual” vs.
“fake-serialized” forms.

Reply:
It’s worth noting also that it’d be sound to key the cache against the
`argv` and the `kwargs` separately. So `configure(["tb", "--logdir=x"])`
and `configure(["tb"], logdir="x")` would map to different cache keys
(again, sound, but incomplete).

Reply:
Keying the cache against them separately seems fine, given that the goal is just to detect when the exact same execution is performed twice, not to dedupe logically equivalent ones (e.g. we don't handle argument re-orders, etc.).

Not a big fan of feeding in the kwargs to argument parsing as extra "argv" entries; it's messier and then you have to ensure all the kwarg values can be be roundtripped through a string conversion correctly, which would require some effort if we have e.g. list or dict-valued arguments in the future.

Reply:
Oh, just saw this. Glad that we’re on the same page. :-)